cmd/vic-machine/common/proxy_test.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2018 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestProcessProxies(t *testing.T) {
+	urls := [...]string{
```

Comment:
Can we put the good urls and bad urls into separate slices so that we can test combinations where `HTTPProxy != HTTPSProxy`? Something like this would also allow us to check combinations more combinations, including combinations of valid & invalid proxies.

This also makes it easier to add new URLs without needing to update the logic later in the test.

I think it also makes sense to replace the example domains suggested by George with domain names that are reserved for such use, to avoid using potentially valid domains that we do not own.

Lastly, I think it's helpful to use the optional `msg` arguments to ensure that a helpful message is printed if the test fails. (Debugging failures in CI can be annoying if all that's logged is something like `true != false`.)

Example of what I mean:

```go
	gurls := [...]string{
		"https://fully.qualified.example.com",
		"https://fully.qualified.example.com:443",
		"http://fully.qualified.example.com",
		"http://fully.qualified.example.com:80", 
		"http://203.0.113.123", 
		"http://[2001:DB8:0123::]", 
	}

	burls := [...]string{
		"example.com",
		"example.com:80",
		"localhost",
		"localhost:80",
		"ftp://example.com",
		"httpd://example.com",
	}

	for _, ghttp := range gproxies {
		for _, ghttps := range gproxies {
			gproxy := Proxies{HTTPProxy: &ghttp, HTTPSProxy: &ghttps}

			_, _, err := gproxy.ProcessProxies()
			assert.NoError(t, err, "Expected %s and %s to be accepted", ghttp, ghttps)
			assert.True(t, gproxy.IsSet, "Expected proxy to be marked as set")
		}
 	} 

	for _, ghttp := range gproxies {
		for _, bhttps := range gproxies {
			bproxy := Proxies{HTTPProxy: &ghttp, HTTPSProxy: &bhttps}

			_, _, err := bproxy.ProcessProxies()
			assert.Error(t, err, "Expected %s to be rejected", bhttps)
		}
 	} 

	for _, bhttp := range gproxies {
		for _, ghttps := range gproxies {
			bproxy := Proxies{HTTPProxy: &bhttp, HTTPSProxy: &ghttps}

			_, _, err := bproxy.ProcessProxies()
			assert.Error(t, err, "Expected %s to be rejected", bhttp)
		}
 	} 

	for _, bhttp := range gproxies {
		for _, bhttps := range gproxies {
			bproxy := Proxies{HTTPProxy: &bhttp, HTTPSProxy: &bhttps}

			_, _, err := bproxy.ProcessProxies()
			assert.Error(t, err, "Expected %s and %s to be rejected", bhttp, bhttps)
		}
 	} 
```

Reply:
Thanks for the suggestions. Initially, I thought most users will set http/https proxy to the same proxy server, and this refactoring definitely covers all cases and easier to extend.